File: 2021/08/03/TwoSum.java
/**
 * https://leetcode.com/problems/two-sum-ii-input-array-is-sorted/
 * @author Andrew Jarombek
 * @since 8/3/2021
 */

public class TwoSum {
    public int[] twoSum(int[] numbers, int target) {

        for (int i = 0; i < numbers.length; i++) {
            for (int j = i + 1; j < numbers.length; j++) {
                int sum = numbers[i] + numbers[j];
                if (sum == target) {
                    return new int[] {i+1, j+1};
                }

                if (sum > target) {
                    break;
                }
            }
        }

        return new int[] {};
    }

    public int[] twoSumSolutionTwo(int[] numbers, int target) {
        int start = 0;
        int end = numbers.length - 1;

        while (start != end) {
            int sum = numbers[start] + numbers[end];
            if (sum == target) {
                return new int[] {start+1, end+1};
            } else if (sum > target) {
                end--;
            } else {
                start++;
            }
        }

        return new int[] {};
    }
}

File: 2021/08/22/twoSum.c
// Program: https://leetcode.com/problems/two-sum/
// Author: Andrew Jarombek
// Date: 8/22/2021

#include <stdlib.h>
#include <assert.h>

int* twoSum(int* nums, int numsSize, int target, int* returnSize) {
    returnSize = malloc(2 * sizeof(int));

    for (int i = 0; i < numsSize; i++) {
        for (int j = 1; j < numsSize; j++) {
            if (nums[i] + nums[j] == target) {
                returnSize[0] = i;
                returnSize[1] = j;
                return returnSize;
            }
        }
    }

    return returnSize;
}

int main() {
    int* nums = malloc(4 * sizeof(int));
    nums[0] = 2;
    nums[1] = 7;
    nums[2] = 11;
    nums[3] = 15;

    int* result = twoSum(nums, 4, 18, 0);
    assert(result[0] == 1);
    assert(result[1] == 2);

    free(nums);
    free(result);
}
